### ISSUE TITLE
build failed on OpenBSD

### ISSUE THREAD

**Issue by semarie** [rust-lang/stacker#32]
while building rustc nightly, the build of stacker failed with the following error:

```
error: unexpected end of macro invocation
   --> /data/semarie/build-rust/build_dir/rustc-nightly-src/vendor/stacker/src/lib.rs:375:95
    |
375 |             assert_eq!(libc::pthread_stackseg_np(libc::pthread_self(), stackinfo.as_mut_ptr()));
    |                                                                                               ^ missing tokens in macro arguments

error[E0308]: mismatched types
   --> /data/semarie/build-rust/build_dir/rustc-nightly-src/vendor/stacker/src/lib.rs:376:18
    |
376 |             Some(stackinfo.assume_init().ss_sp)
    |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `usize`, found *-ptr
    |
    = note:     expected type `usize`
            found raw pointer `*mut libc::c_void`

error: aborting due to 2 previous errors
```

For the first one, I am unsure about the error itself, but I doubt the code will work: the argument passed as stack isn't a valid thread stack.

For the second error, `ss_sp` is a pointer, and not a `usize`. I am also unsure how to properly convert it.

**Comment by nagisa** [comment]
Thanks! Will look into it over the weekend, although you are welcome to send a patch as well.

**Comment by nagisa** [comment]
While I have you here, any suggestions on what would be the easiest way to set up a cross-building for openbsd with github actions?

**Comment by semarie** [comment]
ok, so regarding the errors, having building code will be simple, but working code will not.

- `assert_eq!()` has a missing parameter (`0`)
- and using `as usize` seems working to convert the pointer

but `cargo test` is failing:
```
---- foo stdout ----
thread 'foo' panicked at 'attempt to subtract with overflow', src/lib.rs:76:35
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by semarie** [comment]
@nagisa  regarding cross-building for OpenBSD, I am still unsure that it is doable from linux due to linker. it is something that need checking (but as I don't use Linux...)

**Comment by nagisa** [comment]
Well, there’s a pre-built gcc toolchain targetting freebsd that's easy to obtain on ubuntu flavours. I was thinking of something along the lines of a docker image with the toolchain inside of it. /me shrugs.

**Comment by nagisa** [comment]
As for the overflow, it may be the case that we might need to return `stack_info.ss_sp - stack_info.ss_size` in `guess_os_stack_size` to get the base rather than just `ss_sp`.

**Comment by semarie** [comment]
@nagisa yes, I think using `ss_size` will help. But I fear it will not be enough to make it work on OpenBSD.

Some elements regarding OpenBSD and stack usage:
- you are not allowed to use random memory as stack. it *must* be allocated with the specific flag [MAP_STACK](http://man.openbsd.org/mmap.2#MAP_STACK) to make the kernel know it will be used as stack. Else, the process will be killed (uncatchable signal). Currently, it makes psm `test src/lib.rs - on_stack (line 166)` to fail.

- the exact size of the stack is unknown. OpenBSD adds [random sized gap](http://man.openbsd.org/sysctl.2#KERN_STACKGAPRANDOM~2) at top and bottom to ensure a stack-overflow will kill the process (uncatchable signal). Workaround for it is possible, by reducing the "guessed" stack size.

**Comment by nagisa** [comment]
@semarie I have published stacker 0.1.9 to crates.io.

You will still need to submit a PR to rustc to bump the version in the lockfile for this fix to propagate fully.

**Comment by semarie** [comment]
@nagisa thanks for the publication, and for the notification about the PR on rustc side
